### PR TITLE
fix(network): when connecting to peers, ignore invalid ports, and prefer canonical ports

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We're seeing a lot of invalid peers on the Zcash network, we should try to filter them out.

## Solution

Port rejections:
- Reject Flux/ZelCash and misconfigured Zcash peers
- Check before making initial peer connections, and before adding peers to the address book

Port preference:
- When making outbound connections, prefer peers on the canonical Zcash port for the configured network
- Apply this preference to initial peer connections, and ongoing address book peer connections

Related refactors:
- Move peer address validation to its own module
- Add a network parameter to AddressBook and some MetaAddr methods

This should be covered by existing tests, except for peer port preferences.

## Review

Anyone can review this PR, it might be urgent if the network continues to have a lot of Flux peers, and they're getting gossiped in Zcash address books.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

We might want to prefer publicly routable IP addresses, or reject private addresses with a config - #3117

### CODE REVIEW

File: zebra-network/src/meta_addr.rs [hunk 1]
```diff
@@ -529,8 +536,8 @@ impl MetaAddr {
     ///
     /// Since the addresses in the address book are unique, this check can be
     /// used to permanently reject entire [`MetaAddr`]s.
-    pub fn address_is_valid_for_outbound(&self) -> bool {
-        !self.addr.ip().is_unspecified() && self.addr.port() != 0
+    pub fn address_is_valid_for_outbound(&self, network: Network) -> bool {
+        PeerPreference::new(&self.addr, network).is_ok()
     }
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs:541]: Part of a refactor which moves address validity checks to `PeerPreference`.

File: zebra-network/src/meta_addr.rs [hunk 2]
```diff
@@ -866,6 +873,10 @@ impl Ord for MetaAddr {
         // First, try states that are more likely to work
         let more_reliable_state = self.last_connection_state.cmp(&other.last_connection_state);
 
+        // Then, try addresses that are more likely to be valid.
+        // Currently, this prefers addresses with canonical Zcash ports.
+        let more_likely_valid = self.peer_preference().cmp(&other.peer_preference());
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs:878]: Prefer address book peers on canonical addresses.

(Ok sorts before Err, and Preferred sorts before Acceptable.)

File: zebra-network/src/peer_set/initialize.rs [hunk 3]
```diff
@@ -383,36 +383,62 @@ async fn limit_initial_peers(
...
+        match preference {
+            Ok(preference) => preferred_peers
+                .entry(preference)
+                .or_default()
+                .push(peer_addr),
+            Err(error) => warn!(
+                ?peer_addr,
+                ?error,
+                "invalid initial peer from DNS seeder or configured IP address",
+            ),
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs:411]: Sort valid initial peers by preference, and remove invalid peers.

File: zebra-network/src/peer_set/initialize.rs [hunk 4]
```diff
@@ -383,36 +383,62 @@ async fn limit_initial_peers(
...
+    // choosing preferred peers first.
+    let mut initial_peers: HashSet<SocketAddr> = HashSet::new();
+    for better_peers in preferred_peers.values() {
+        let mut better_peers = better_peers.clone();
+        let (chosen_peers, _unused_peers) = better_peers.partial_shuffle(
+            &mut rand::thread_rng(),
+            config.peerset_initial_target_size - initial_peers.len(),
+        );
+
+        initial_peers.extend(chosen_peers.iter());
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs:434]: Use the best initial peers first.

File: zebra-network/src/peer/priority.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    if peer_addr.ip().is_unspecified() {
+        return Err("invalid peer IP address: unspecified addresses can not be used for outbound connections");
+    }
+
+    // 0 is an invalid outbound port.
+    if peer_addr.port() == 0 {
+        return Err(
+            "invalid peer port: unspecified ports can not be used for outbound connections",
+        );
+    }
```
Comment (teor2345) [inline zebra-network/src/peer/priority.rs:102]: These checks have been moved as part of the refactor.

File: zebra-network/src/peer/priority.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            );
+        } else if peer_addr.port() == 18232 {
+            return Err(
+                "invalid peer port: port is for Testnet, but this node is configured for Mainnet",
+            );
+        } else if [16125, 26125].contains(&peer_addr.port()) {
+            // 16125/26125 is used by Flux/ZelCash, which uses the same network magic numbers as Zcash,
+            // so we have to reject it by port
+            return Err("invalid peer port: port is for a non-Zcash network");
+        }
```
Comment (teor2345) [inline zebra-network/src/peer/priority.rs]: Check for misconfigured Zcash and Flux peers.

File: zebra-network/src/peer/priority.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    pub fn new(
+        peer_addr: &SocketAddr,
+        network: impl Into<Option<Network>>,
+    ) -> Result<PeerPreference, &'static str> {
+        address_is_valid_for_outbound_connections(peer_addr, network)?;
+
+        // This check only prefers the configured network,
+        // because the address book and initial peer connections reject the port used by the other network.
+        let canonical_port =
+            AttributePreference::preferred_from([8232, 18232].contains(&peer_addr.port()));
```
Comment (teor2345) [inline zebra-network/src/peer/priority.rs:76]: Prefer peers on the canonical Zcash port.

**Review by teor2345 — COMMENTED** [review]
This PR has a lot of test refactors, so I've added comments describing the actual changes.

File: zebra-network/src/peer/priority.rs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    if let Some(network) = network.into() {
+        if peer_addr.port() == network.default_port() {
+            return Ok(());
+        }
+
+        if peer_addr.port() == 8232 {
+            return Err(
+                "invalid peer port: port is for Mainnet, but this node is configured for Testnet",
+            );
+        } else if peer_addr.port() == 18232 {
```
Comment (teor2345) [inline zebra-network/src/peer/priority.rs:114]: TODO: add 18344 regtest case here:
https://github.com/ZcashFoundation/dnsseeder/blob/8aa4e10c8fbfdfba8af00957ef5d49717432c215/zcash/network/params.go#L14

Comment (teor2345) [reply zebra-network/src/peer/priority.rs:114]: Fixed in [9448698](https://github.com/ZcashFoundation/zebra/pull/4564/commits/9448698bafdf22447318bfa8a0ce285eaef909f8)

File: zebra-network/src/peer/priority.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
+//! Priorising outbound peer connections based on peer attributes.
```
Comment (upbqdn) [inline zebra-network/src/peer/priority.rs]:
```suggestion
//! Prioritizing outbound peer connections based on peer attributes.
```

File: zebra-network/src/peer_set/initialize.rs [hunk 10]
```diff
@@ -383,36 +383,62 @@ async fn limit_initial_peers(
     config: &Config,
     address_book_updater: tokio::sync::mpsc::Sender<MetaAddrChange>,
 ) -> HashSet<SocketAddr> {
-    let all_peers = config.initial_peers().await;
-    let peers_count = all_peers.len();
+    let all_peers: HashSet<SocketAddr> = config.initial_peers().await;
```
Comment (upbqdn) [inline zebra-network/src/peer_set/initialize.rs:386]: I wonder, why did you state the type for `all_peers` explicitly?

File: zebra-network/src/peer_set/initialize.rs [hunk 11]
```diff
@@ -383,36 +383,62 @@ async fn limit_initial_peers(
...
+            Err(error) => warn!(
+                ?peer_addr,
+                ?error,
+                "invalid initial peer from DNS seeder or configured IP address",
+            ),
+        }
+    }
 
-    if all_peers.len() > config.peerset_initial_target_size {
+    let peers_count = preferred_peers.len();
```
Comment (upbqdn) [inline zebra-network/src/peer_set/initialize.rs]: I'm unnecessarily punctual here, but `peers_count` seems like it's associated with `all_peers` instead of `preferred_peers`. Maybe `preferred_peers_count`?

File: zebra-network/src/peer_set/initialize.rs [hunk 12]
```diff
@@ -383,36 +383,62 @@ async fn limit_initial_peers(
...
         // `send` only waits when the channel is full.
         // The address book updater runs in its own thread, so we will only wait for a short time.
         let _ = address_book_updater.send(peer_addr).await;
     }
 
-    initial_peers.iter().copied().collect()
+    // Split out the `initial_peers` that will be shuffled and returned,
+    // choosing preferred peers first.
+    let mut initial_peers: HashSet<SocketAddr> = HashSet::new();
+    for better_peers in preferred_peers.values() {
```
Comment (upbqdn) [inline zebra-network/src/peer_set/initialize.rs:427]: I think this would start working backward if we swapped the order of the enum variants here: https://github.com/ZcashFoundation/zebra/blob/peer-port-priority/zebra-network/src/peer/priority.rs#L13. Are we OK with that? The swap could be made in an unrelated refactor for example.

**Review by upbqdn — DISMISSED** [review]
Looks very good. I left some observations.

File: zebra-network/src/meta_addr.rs (hunk 2 above)
Comment (upbqdn) [reply zebra-network/src/meta_addr.rs:878]: Please check https://github.com/ZcashFoundation/zebra/pull/4564#discussion_r890471136.

File: zebra-network/src/peer_set/initialize.rs (hunk 10 above)
Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs:386]: It's a bit tricky to convert a `HashSet<SocketAddr>` into a `BTreeMap<PeerPreference, Vec<SocketAddr>>`.

So I added both types to the code, to make it clear what was happening.

File: zebra-network/src/peer_set/initialize.rs (hunk 11 above)
Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs]: That's actually a bug - we need to log the total number of peers (`all_peers.len()`), not the number of active preference categories (`preferred_peers.len()`).

File: zebra-network/src/peer/priority.rs (hunk 9 above)
Comment (teor2345) [reply zebra-network/src/peer/priority.rs]: Fixed in 3e11db0b9

File: zebra-network/src/peer_set/initialize.rs (hunk 11 above)
Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs]: Fixed in 5003893a2

File: zebra-network/src/peer_set/initialize.rs (hunk 12 above)
Comment (teor2345) [reply zebra-network/src/peer_set/initialize.rs:427]: Sounds risky, I documented where order is required in 3e11db0b9

**Review by upbqdn — APPROVED** [review]
It's great how you can split all changes into commits from the very beginning.

**Comment by teor2345** [comment]
> It's great how you can split all changes into commits from the very beginning.

Sometimes it works better than others!
I also use `git commit --fixup <commit hash>` a lot.

We can chat about how to plan out a PR some time if you'd like.

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
This is an intermittent macOS error, we might want to disable this test on that OS if it continues:
> test zebra_rpc_conflict has been running for over 60 seconds

2h5m later:
> Error: The operation was canceled.

https://github.com/ZcashFoundation/zebra/runs/6870414021?check_suite_focus=true#step:13:54470

**Comment by teor2345** [comment]
@mergifyio update
